src/window.c

```diff
@@ -4416,6 +4418,13 @@ close_others(
     for (wp = firstwin; win_valid(wp); wp = nextwp)
     {
 	nextwp = wp->w_next;
+
+	// autocommands messed this one up
+	if (old_curwin != curwin && win_valid(old_curwin))
+	    curwin = old_curwin;
+	if (old_curbuf != curwin->w_buffer && buf_valid(old_curbuf))
+	    curwin->w_buffer = old_curbuf;
```

Comment:
I think doing this without changing `curbuf->b_nwindows` and `old_curbuf->b_nwindows` can cause problems as well. Simply using `curbuf = curwin->w_buffer` is probably safer.